### ISSUE TITLE
Use torch.einsum for performance gains

### ISSUE THREAD

**Issue by LiamConnell** [LiamConnell/FootyAI#5]
### 🔄 Refactor to `torch.einsum` where it buys us something

Use Einstein summation to fuse “multiply + reduce” patterns, cut temporary tensors, and (usually) hit a faster kernel. Below are the spots that are both *clearer* and *worthwhile* to change.

---

#### ☑️ 1. Kick aggregation in `TorchSoccerEnv.step`

File `src/torch_soccer_env.py`

* **Replace**

  ```python
  team1_kick_mask  = team1_can_kick.float().unsqueeze(-1)      # (B, P, 1)
  team1_total_kicks = torch.sum(team1_kick_mask * team1_kick_vel, dim=1)
  team1_kickers_count = torch.sum(team1_kick_mask, dim=(1, 2))
  # same for team 2
  ```
* **With**

  ```python
  m1 = team1_can_kick.float()                   # (B, P)
  v1 = team1_kick_vel                          # (B, P, 2)
  team1_total_kicks   = torch.einsum('bp,bpj->bj', m1, v1)
  team1_kickers_count = torch.einsum('bp->b',          m1)

  m2 = team2_can_kick.float()
  v2 = team2_kick_vel
  team2_total_kicks   = torch.einsum('bp,bpj->bj', m2, v2)
  team2_kickers_count = torch.einsum('bp->b',          m2)
  ```

---

#### ☑️ 2. Policy‑gradient loss in `update_policy`

File `src/train_self_play.py`

* **Replace**

  ```python
  loss_a = - (log_probs_a.sum(dim=-1) * discounted_rewards_a).sum(dim=0).mean()
  loss_b = - (log_probs_b.sum(dim=-1) * discounted_rewards_b).sum(dim=0).mean()
  ```
* **With**

  ```python
  B = log_probs_a.shape[1]  # batch
  loss_a = -torch.einsum('tba,tb->', log_probs_a,  discounted_rewards_a) / B
  loss_b = -torch.einsum('tba,tb->', log_probs_b,  discounted_rewards_b) / B
  ```

---

#### ☐ 3. ( Optional ) Squared distances in `TorchSoccerEnv.reset/step`

If we ever need distances **without** the expensive `sqrt`, we can fuse the dot‑product:

```python
diff = self.team1_positions - self.ball_position.unsqueeze(1)   # (B, P, 2)
team1_sq_dist = torch.einsum('bpj,bpj->bp', diff, diff)
```

Not urgent; current `torch.norm` is already fine.

---

### Acceptance criteria

* [ ] All targeted sites swapped to the `einsum` versions above.
* [ ] Unit tests / training runs still pass.
* [ ] Benchmarks show equal or better wall‑time for the affected kernels.

> *Rationale*: These changes eliminate explicit broadcasting + multiple reductions, replacing them with a single contraction that PyTorch maps to `baddbmm`/BLAS under the hood, reducing memory traffic and clarifying intent.
